File: ios-src/DownloadManagerKit/Sources/DownloadManagerKit/DownloadItem.swift
//
//  DownloadItem.swift
//  DownloadManagerKit
//

import Foundation

/// A value type that represents an item to be downloaded.
/// Used to track the status and progress of a download operation.
public struct DownloadItem: Identifiable, Codable, Sendable {
   public var id: URL { path }
   
   public let url: URL
   public let path: URL
   public private(set) var progress: Double
   public private(set) var status: DownloadStatus
   public var resumeDataPath: URL?
   
   init(url: URL, path: URL, progress: Double = 0.0, status: DownloadStatus = .idle, resumeDataPath: URL? = nil) {
      self.url = url
      self.path = path
      self.progress = progress
      self.status = status
      self.resumeDataPath = resumeDataPath
   }
   
   public mutating func setProgress(_ progress: Double) {
      self.progress = progress
   }
   
   public mutating func setStatus(_ status: DownloadStatus) {
      self.status = status
   }
}

File: android/lib/src/main/java/org/silvermine/downloadmanager/DownloadItem.kt
package org.silvermine.downloadmanager

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

/**
 * A value type that represents an item to be downloaded.
 * Used to track the status and progress of a download operation.
 */
@Serializable
data class DownloadItem(
   @SerialName("url")
   val url: String,

   @SerialName("path")
   val path: String,

   @SerialName("progress")
   val progress: Double = 0.0,

   @SerialName("transferredBytes")
   val transferredBytes: Long = 0,

   @SerialName("totalBytes")
   val totalBytes: Long? = null,

   @SerialName("status")
   val status: DownloadStatus = DownloadStatus.Idle,
) {
   fun withTransfer(newTransferredBytes: Long, newTotalBytes: Long?): DownloadItem =
      copy(
         progress = derivedProgress(newTransferredBytes, newTotalBytes),
         transferredBytes = newTransferredBytes,
         totalBytes = newTotalBytes,
      )

   fun withStatus(newStatus: DownloadStatus): DownloadItem =
      copy(status = newStatus).let { updated ->
         val updatedTotalBytes = if (newStatus == DownloadStatus.Completed) {
            updated.totalBytes ?: updated.transferredBytes
         } else {
            updated.totalBytes
         }

         updated.copy(
            progress = derivedProgress(updated.transferredBytes, updatedTotalBytes, newStatus),
            totalBytes = updatedTotalBytes,
         )
      }

   private fun derivedProgress(
      transferredBytes: Long,
      totalBytes: Long?,
      currentStatus: DownloadStatus = status,
   ): Double =
      when {
         currentStatus == DownloadStatus.Completed -> 100.0
         totalBytes != null && totalBytes > 0 -> {
            (transferredBytes.toDouble() / totalBytes.toDouble()) * 100.0
         }
         else -> 0.0
      }
}
